client/gutenberg/editor/async-loader.jsx

```diff
@@ -0,0 +1,60 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+import { Component } from 'react';
+
+export const asyncLoader = ( { promises, loading, success, failure } ) =>
+	class AsyncLoader extends Component {
+		state = {
+			results: Object.keys( promises ).reduce(
+				( output, key ) => ( { ...output, [ key ]: null } ),
+				{}
+			),
+		};
+
+		componentDidMount() {
+			// here we want to make sure all promises "resolve" even when
+			// they reject because Promise.all will return a single value
+			// on the first rejection instead of an array as when they all
+			// resolve. this is for our own accounting and ensures that
+			// we wait for all promises to fulfill
+			const runners = Object.keys( promises ).map( key =>
+				promises[ key ]
+					.then( a => {
+						this.setState( state => ( { results: { ...( state || {} ).results, [ key ]: a } } ) );
+
+						return a;
+					} )
+					.then( a => [ true, a ], a => [ false, a ] )
+			);
+
+			Promise.all( runners ).then( resolutions => {
+				const keys = Object.keys( promises );
+				const [ results, successful ] = resolutions.reduce(
+					( [ output, allSuccess ], [ wasSuccess, data ], index ) => [
+						{ ...output, [ keys[ index ] ]: data },
+						allSuccess && wasSuccess,
+					],
+					[ {}, true ]
+				);
+
+				this.setState( { results, successful } );
+			} );
+		}
+
```

Comment:
Might add a `componentDidUpdate` handler that yells if the promises change? I expect we'd force the user of this component to use a `key` to prevent that kind of update.

Reply:
`promises` isn't a prop tho but an argument to the surrounding factory. So I don't think we can use `componentDidUpdate` to guard against that -- but on the plus side, that should also mean that we plainly don't have to worry it might change.